pypesto/result/sample.py

```diff
@@ -5,19 +5,113 @@
 import numpy as np
 
 
-class McmcPtResult(dict):
+class SampleResultBase(dict):
+    """
+    Sampling result object base.
+
+    Parameters
+    ----------
+    trace_x: [n_samples, n_par]
+        Parameters.
+        (Caution: indexing is different for McmcPtResult)
+    trace_neglogpost: [n_samples]
+        Negative log posterior values.
+        (Caution: indexing is different for McmcPtResult)
+    trace_neglogprior: [n_samples]
+        Negative log prior values.
+        (Caution: indexing is different for McmcPtResult)
+    burn_in: int
+        The burn in index.
+    time: float
+        The computation time.
+    auto_correlation:
+        The estimated autcorrelation.
+    effective_sample_size:
+        The estimated effective sample size.
+    message: str
+        Textual comment on the profile result.
+    """
+
+    def __init__(
+        self,
+        trace_x: np.ndarray = None,
+        trace_neglogpost: np.ndarray = None,
+        trace_neglogprior: np.ndarray = None,
+        burn_in: int = None,
+        time: float = 0.0,
+        auto_correlation: float = None,
+        effective_sample_size: float = None,
+        message: str = None,
+        **kwargs,
+    ):
+
+        super().__init__()
+
+        self.trace_x = trace_x
+        self.trace_neglogpost = trace_neglogpost
+        self.trace_neglogprior = trace_neglogprior
+        self.burn_in = burn_in
+        self.time = time
+        self.auto_correlation = auto_correlation
+        self.effective_sample_size = effective_sample_size
+        self.message = message
+        self._index_iter = None
+
+    def _check_trace_dimensions(self):
+        raise NotImplementedError(
+            'Base class does not implement trace checks.'
+        )
+
+    @property
+    def n_samples(self):
+        """Return number of samples."""
+        return 0 if (self.trace_x is None) else self.trace_x.shape[0]
+
+    def __getattr__(self, key):
+        """Allow usage of keys like attributes."""
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(key)
+
+    def __setattr__(self, key, value):
+        """Allow usage of keys like attributes."""
+        if key == 'n_samples':
+            raise ValueError(
+                "n_samples cannot be set, as they are computed "
+                "from self.trace_x."
+            )
+        self[key] = value
+
+    def __iter__(self):
```

Comment:
would remove `__iter__` and `__next__` as I don't see the special usecase of iterating over x, but not fval, and using one x at a time (in contrast to slicing to a single parameter, e.g. for plotting, which would also be done via `trace_x[:, ix]`, but also fine to leave if preferred.